From 3f82e7ec9cca04858e506a0c46ee2f4b37445cf0 Mon Sep 17 00:00:00 2001
From: Pardha Kesapragada <pkesapragada@habana.ai>
Date: Tue, 23 Mar 2021 15:46:48 +0200
Subject: [PATCH] [SW-37240] add provision to clear sw status registers

During warm/prstn reset, SW status registers reflect
old values (from a last successful execution) as these
registers are reset only by the cold reset. Hence they
have to be manually cleared to show current execution
status only.

Change-Id: I5e3b9ade76444959e7e720f421aa6dedf25f0324
Signed-off-by: Pardha Kesapragada <pkesapragada@habana.ai>
---

diff --git a/app/arcboot_rom.c b/app/arcboot_rom.c
index 00ef437..d92e3c7 100644
--- a/app/arcboot_rom.c
+++ b/app/arcboot_rom.c
@@ -128,6 +128,9 @@
 	/* lock everything as soon as possible */
 	security_disable_interfaces();
 
+	/* clear SW Status registers */
+	arcboot_rom_clear_sw_status();
+
 	arcboot_rom_status_update(PROGRESS_INIT);
 
 	/* make sure spi controller doesn't swap bytes */
diff --git a/common/common.h b/common/common.h
index aa62007..ce940ce 100644
--- a/common/common.h
+++ b/common/common.h
@@ -13,5 +13,6 @@
 void update_warm_reset_config(void);
 void arcboot_rom_handle_tpm(void);
 void arcboot_rom_request_cpld_wdog(void);
+void arcboot_rom_clear_sw_status(void);
 
 #endif /* _COMMON_H_ */
diff --git a/gaudi2_arc_gnu/common.c b/gaudi2_arc_gnu/common.c
index d01465b..6a6e784 100644
--- a/gaudi2_arc_gnu/common.c
+++ b/gaudi2_arc_gnu/common.c
@@ -487,3 +487,18 @@
 	val &= ~mask;
 	regmap_writel(val, mmPSOC_GPIO3_DATAOUT);
 }
+
+void arcboot_rom_clear_sw_status(void)
+{
+	/* As per Architecture recommendation, clear status reg */
+	regmap_writel(0, mmPSOC_SECURITY_STATUS);
+
+	regmap_writel(0, mmPSOC_SECURITY_SW_STATUS_0);
+	regmap_writel(0, mmPSOC_SECURITY_SW_STATUS_1);
+	regmap_writel(0, mmPSOC_SECURITY_SW_STATUS_2);
+	regmap_writel(0, mmPSOC_SECURITY_SW_STATUS_3);
+	regmap_writel(0, mmPSOC_SECURITY_SW_STATUS_4);
+	regmap_writel(0, mmPSOC_SECURITY_SW_STATUS_5);
+	regmap_writel(0, mmPSOC_SECURITY_SW_STATUS_6);
+	regmap_writel(0, mmPSOC_SECURITY_SW_STATUS_7);
+}
diff --git a/greco_arc/common.c b/greco_arc/common.c
index aeeed80..b409f6b 100644
--- a/greco_arc/common.c
+++ b/greco_arc/common.c
@@ -172,3 +172,8 @@
 {
 	/* Do this after boot rom code, since it is beyond TO */
 }
+
+void arcboot_rom_clear_sw_status(void)
+{
+	/* Do this after boot rom code, since it is beyond TO */
+}
